opentelemetry-user-events-trace/src/traces/exporter.rs

```diff
@@ -0,0 +1,175 @@
+
+/// UserEventsSpanExporter exports spans in EventHeader format to user_events tracepoint.
+pub(crate) struct UserEventsSpanExporter {
+    provider: Mutex<Provider>,
+    name: String,
+    event_sets: Vec<Arc<EventSet>>,
+}
+
+impl Debug for UserEventsSpanExporter {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "user_events log exporter (provider: {})", self.name)
+    }
+}
+
+use futures_util::future::BoxFuture;
+use opentelemetry_sdk::trace::SpanExporter;
+
+impl SpanExporter for UserEventsSpanExporter {
+    fn export(&mut self, batch: Vec<SpanData>) -> BoxFuture<'static, OTelSdkResult> {
+        let result = batch.iter().try_for_each(|span| self.export_span(span));
+
+        Box::pin(async move { result })
+    }
+
+    fn shutdown(&mut self) -> OTelSdkResult {
+        // The explicit unregister() is done in shutdown()
+        // as it may not be possible to unregister during Drop
+        // as Loggers are typically *not* dropped.
+        if let Ok(mut provider) = self.provider.lock() {
+            provider.unregister();
+            Ok(())
+        } else {
+            Err(OTelSdkError::InternalFailure(
+                "Failed to acquire lock on provider".to_string(),
+            ))
+        }
+    }
+}
+
+impl UserEventsSpanExporter {
+    /// Create a new instance of the exporter
+    pub(crate) fn new(provider_name: &str) -> Result<Self, String> {
+        if provider_name.len() >= 234 {
+            return Err("Provider name must be less than 234 characters.".to_string());
+        }
+        if !provider_name
+            .chars()
+            .all(|c| c.is_ascii_alphanumeric() || c == '_')
+        {
+            return Err(
+                "Provider name must contain only ASCII letters, digits, and '_'.".to_string(),
+            );
+        }
+
+        let mut eventheader_provider = Provider::new(provider_name, &Provider::new_options());
+        let event_sets = Self::register_events(&mut eventheader_provider);
+        let name = eventheader_provider.name().to_string();
+
+        Ok(UserEventsSpanExporter {
+            provider: Mutex::new(eventheader_provider),
+            name,
+            event_sets,
+        })
+    }
+
+    fn register_events(eventheader_provider: &mut Provider) -> Vec<Arc<EventSet>> {
+        let keyword: u64 = 1;
+        let levels = [
+            Level::CriticalError,
+            Level::Error,
+            Level::Warning,
+            Level::Informational,
+            Level::Verbose,
+        ];
+
+        let mut event_sets = vec![Arc::new(EventSet::new_unregistered())];
+
+        for &level in levels.iter() {
+            let event_set = eventheader_provider.register_set(level, keyword);
+            event_sets.push(event_set);
+        }
+        event_sets
+    }
+
+    pub(crate) fn export_span(&self, span: &SpanData) -> OTelSdkResult {
+        let level = Level::Informational; //TODO which level/tracepoint to use?
```

Comment:
I suggest to hardcode information and use that. Create the tracepoint in new(), and use it here. It should never fail as we create it in new() anyway.

Reply:
Didn't get that. Do you mean creating only single tracepoint (say) providename_L4K1, and use that?

Reply:
yes.